### PR TITLE
feat: add start_of_part directive

### LINKED ISSUE #821: 💡 RFC: Start Of Part directive
### Background & Motivation

implement {sop} {start_of_part}

### PR DESCRIPTION
**Pull request by isaiahdahl**
fixes #821

### CODE REVIEW

**Comment by isaiahdahl** [comment]
@martijnversluis  Interesting thought with the start_of_part directive that I had. What if we allowed a short form for start_of_part to be `p` (as well as `sop` and `eop`. Could also add `ep` for another short of end of part but the reason for this thought is because the "part" type is identifying the paragraph type from the first word lowercased. you could have an entire sheet just use the {p: syntax exactly like it's so commonly used with {c: syntax (where there's no closing directive) but it would be able to support all the custom formatting rules per section type because it's still going to call that paragraph a "chorus" or "verse" 

so 

```
{p: Intro (2x)}
[Gm][F][/][/][|][C#dim7][Dm][/][/][|]

{p: Verse 1}
My [Dm7]heart has always [C/E]longed for something[F] more

{p: Chorus 1 *2}
Thine is the [Gm]King - [F]dom, the power, the [C#dim7]glo - [Dm]ry

```

**Comment by isaiahdahl** [comment]
@martijnversluis I see this is now rebased from a fresh pull from upstream? Should I go and fix the test failures, were there any other comments?

**Comment by martijnversluis** [comment]
Yeah, I had to rebase because I extracted the song building to a separate builder class. Did those failures also show before the rebase?

File: src/song_builder.ts [hunk 1]
```diff
@@ -189,8 +189,14 @@ class SongBuilder {
 
   startSection(sectionType: ParagraphType, tag: Tag): void {
     this.checkCurrentSectionType(NONE, tag);
-    this.sectionType = sectionType;
-    this.setCurrentProperties(sectionType);
+
+    if (sectionType === PART && tag.value) {
+      this.sectionType = tag.value.split(' ')[0].toLowerCase();
+    } else {
+      this.sectionType = sectionType;
```
Comment (martijnversluis) [inline src/song_builder.ts:196]: Maybe we should add a warning for a tag without value, and ignore it?

**Comment by martijnversluis** [comment]
> @martijnversluis I see this is now rebased from a fresh pull from upstream? Should I go and fix the test failures, were there any other comments?

I added one other comment, seems like a great addition! 👍

File: src/song_builder.ts (hunk 1 above)
Comment (isaiahdahl) [reply src/song_builder.ts:196]: This specifically handled the case where the opening is {sop: Tag} and the closing tag is {eop} where the sectionType is  the value and the incoming tag value is null. 

I feel like it would get perfectly acceptable to have a sop eop without values and it would just get processed as a sectionType 'part'

Thoughts?

**Comment by isaiahdahl** [comment]
Yea the tests were all passing previously and there was just a weird edge case in one of the nodejs versions. 

Should be pretty easy to fix but just wasn't sure if there was anything else you saw when rebasing.

File: src/song_builder.ts (hunk 1 above)
Comment (isaiahdahl) [reply src/song_builder.ts:196]: actually sorry I thought this comment was actually for this part

```
checkCurrentSectionType(sectionType: ParagraphType, tag: Tag): void {
    if (this.sectionType !== sectionType && !(sectionType === 'part' && tag.name === 'end_of_part')) {
      this.addWarning(`Unexpected tag {${tag.originalName}}, current section is: ${this.sectionType}`, tag);
    }
  }
```

ignore the first part of my comment. 

But I still think having sop without value should be valid with no warnings.

**Comment by isaiahdahl** [comment]
@martijnversluis thinking since this is based from master on this branch which has the pdf stuff in it and that's not ready to go upstream once these tests pass I'll merge it into master here but then after it's merged into master in this repo we/you could rebase this branch from upstream master and then make a PR just for this feature up to the main one?
